src/Storage/Builder.php

```diff
@@ -0,0 +1,231 @@
+
+    /**
+     * The token signature
+     *
+     * @var string|null
+     */
+    private $signature;
+
+    /**
+     * The data encoder
+     *
+     * @var Parsing\Encoder
+     */
+    private $encoder;
+
+    /**
+     * Initializes a new builder
+     */
+    public function __construct(Parsing\Encoder $encoder)
+    {
+        $this->encoder = $encoder;
+        $this->headers = ['typ'=> 'JWT', 'alg' => 'none'];
+        $this->claims = [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function canOnlyBeUsedBy(string $audience, bool $addHeader = false): BuilderInterface
+    {
+        $audiences = $this->claims['aud'] ?? [];
+        $audiences[] = $audience;
+
+        return $this->setRegisteredClaim(
+            'aud',
+            array_values(array_map('strval', $audiences)),
+            $addHeader
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function expiresAt(int $expiration, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('exp', $expiration, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function withId(string $id, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('jti', $id, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function issuedAt(int $issuedAt, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('iat', (int) $issuedAt, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function issuedBy(string $issuer, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('iss', $issuer, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function canOnlyBeUsedAfter(int $notBefore, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('nbf', $notBefore, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function relatedTo(string $subject, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('sub', $subject, $addHeader);
+    }
+
+    /**
+     * Configures a registered claim
+     */
+    private function setRegisteredClaim(string $name, $value, bool $addHeader): BuilderInterface
+    {
+        $this->with($name, $value);
+
+        if ($addHeader) {
+            $this->headers[$name] = $this->claims[$name];
+        }
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function withHeader(string $name, $value): BuilderInterface
+    {
+        if ($this->signature !== null) {
+            throw new BadMethodCallException('You must unsign before make changes');
+        }
+
+        $this->headers[$name] = $value;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function with(string $name, $value): BuilderInterface
+    {
+        if ($this->signature !== null) {
```

Comment:
Signing the token on a late stage will remove the need of having that exception, which is good.